---
title: "Example workflow of scABC"
author: "Timothy Daley"
date: "4/30/2017"
output: html_document
---

We're going to walk through an example workflow for the scABC package, walking through all the individual steps of the analysis.  We'll use the 6 cell line mixture for an example.  The 1632 cells can be downloaded from GEO under accession number GSE65360 (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE65360).  

## Pre-processing

We used Kundaje pipeline for aligning pair ended reads to hg19 and removing duplicates (https://github.com/kundajelab/atac_dnase_pipelines) with the command below for each read pair file.
```{r engine='bash', eval=FALSE}
bds_scr [SCR_NAME] atac.bds -align -species hg19 -species_file [SPECIES_FILE_PATH] -nth [NUM_THREADS] -fastq1_1 [READ_PAIR1] -fastq1_2 [READ_PAIR2]
```
For details on the pipeline, visit the Kundaje lab website or github page.

The resulting bam files were merged using samtools.
```{r engine='bash', eval=FALSE}
samtools merge [AGGREGATE_BAM] *.trim.PE2SE.nodup.bam
```
The merged bam file was then used as input into MACS2 to call merged peaks for later analysis, using the Kundaje pipeline again.
```{r engine='bash', eval=FALSE}
bds_scr [SCR_NAME] atac.bds -species hg19 -species_file [SPECIES_FILE_PATH] -nth [NUM_THREADS] -se -filt_bam [AGGREGATE_BAM]
```

Several of the cells have multiple runs.  We merged these runs using samtools.

## Obtaining and filtering counts

We will use scABC to analyse the single cell data.  The cell level bam files and their samtools indexes are contained in the folder bams.
```{r cache=TRUE}
setwd("~/scRNAseqAnalysis/scATAC-RNAseq/scABC/vignettes/")
source("~/scRNAseqAnalysis/scATAC-RNAseq/scABC/R/read_data.R")
library(GenomicRanges)
library(Rsamtools)
bamfile.table = read.table(file = "NoTreatment/6Lines/SRX&Type&Batch.txt")
head(bamfile.table)
bamfiles = paste0("bams/", bamfile.table[,1], ".bam")
peaks = select_peaks("NoTreatment/6Lines/mergeAll.tn5.pf.gappedPeak")
dim(peaks)
ForeGround = get_counts_matrix(bamfiles, peaks)
ForeGroundFiltered = filter_peaks(ForeGround$ForeGround, ForeGround$peaks)
peaks = ForeGroundFiltered$peaks
BackGround = get_background(bamfiles, ForeGroundFiltered$peaks)
ForeGroundBackGroundFiltered = filter_background(ForeGround = ForeGroundFiltered$ForeGround, BackGround = BackGround$BackGround)
```
```{r cache=TRUE, echo=FALSE}
#ForeGround=read.table(file = "6cellInSilicoMixtureFullForeGroundPeakFiltered.txt", header=TRUE)
```
```{r}
#ForeGround[1:6, 1:6]
```