---
title: "Example workflow of scABC"
author: "Timothy Daley"
date: "4/30/2017"
output:
  html_document: default
  pdf_document: default
---

We're going to walk through an example workflow for the scABC package, walking through all the individual steps of the analysis.  We'll use the 6 cell line mixture for an example.  The 1632 cells can be downloaded from GEO under accession number GSE65360 (https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE65360).  

## Pre-processing

We used Kundaje pipeline for aligning pair ended reads to hg19 and removing duplicates (https://github.com/kundajelab/atac_dnase_pipelines) with the command below for each read pair file.
```{r engine='bash', eval=FALSE}
bds_scr [SCR_NAME] atac.bds -align -species hg19 -species_file [SPECIES_FILE_PATH] -nth [NUM_THREADS] -fastq1_1 [READ_PAIR1] -fastq1_2 [READ_PAIR2]
```
For details on the pipeline, visit the Kundaje lab website or github page.

The resulting bam files were merged using samtools.
```{r engine='bash', eval=FALSE}
samtools merge [AGGREGATE_BAM] *.trim.PE2SE.nodup.bam
```
The merged bam file was then used as input into MACS2 to call merged peaks for later analysis, using the Kundaje pipeline again.
```{r engine='bash', eval=FALSE}
bds_scr [SCR_NAME] atac.bds -species hg19 -species_file [SPECIES_FILE_PATH] -nth [NUM_THREADS] -se -filt_bam [AGGREGATE_BAM]
```

Several of the cells have multiple runs.  We merged these runs using samtools.

## Obtaining and filtering counts

We will use scABC to analyse the single cell data.  The cell level bam files and their samtools indexes are contained in the folder bams.

```{r}
# always run
setwd("~/scRNAseqAnalysis/scATAC-RNAseq/scABC/vignettes/")
library(devtools)
devtools::install_github("timydaley/scABC", force = TRUE)
library(scABC)
# table that has experiment info
bamfile.table = read.table(file = "NoTreatment/6Lines/SRX&Type&Batch.txt")
head(bamfile.table)
bamfiles = paste0("bams/", bamfile.table[,1], ".bam")
peaks = selectPeaks("NoTreatment/6Lines/mergeAll.tn5.pf.gappedPeak")
dim(peaks)
```

Computing the Forground and Background matrix takes quite a bit of time.  We can do this once and cache the results for later.

```{r cache=TRUE}
#
ForeGround = getCountsMatrix(bamfiles, peaks);
```

```{r cache=TRUE}
#
ForeGroundFiltered = filterPeaks(ForeGround$ForeGroundMatrix, peaks)
peaks = ForeGroundFiltered$peaks
```

```{r cache=TRUE}
#
BackGround = getBackground(bamfiles, peaks)
ForeGroundBackGroundFiltered = filterSamples(ForeGround = ForeGroundFiltered$ForeGroundMatrix, BackGround = BackGround$BackGroundMatrix)
InSilicoForeGround = ForeGroundBackGroundFiltered$ForeGroundMatrix
InSilicoBackGround = ForeGroundBackGroundFiltered$BackGroundMatrix
InSilicoPeaks = peaks
```

Now let's compute the landmarks and take a look at them.

```{r message=FALSE, cache=TRUE}
#
InSilicoLandMarks = computeLandmarks(ForeGround = InSilicoForeGround, 
                                      BackGround = InSilicoBackGround, 
                                      nCluster = 6, lambda = 1, nTop = 2000)
# look at correlation between landmarks
cor(InSilicoLandMarks, InSilicoLandMarks, method = 'spearman')
```

It looks like the cluster landmarks are well seperated.  Let's look at the cell to cluster assignments.

```{r cache=TRUE}
# assign cells to the closest landmark #
InSilicoLandMarkAssignments = assign2landmarks(InSilicoForeGround, InSilicoLandMarks)
InSilicoCell2LandmarkCorrelation = cbind(apply(InSilicoForeGround, 2, function(x) cor(x, InSilicoLandMarks[,1], method = 'spearman')), 
                                         apply(InSilicoForeGround, 2, function(x) cor(x, InSilicoLandMarks[,2], method = 'spearman')), 
                                         apply(InSilicoForeGround, 2, function(x) cor(x, InSilicoLandMarks[,3], method = 'spearman')), 
                                         apply(InSilicoForeGround, 2, function(x) cor(x, InSilicoLandMarks[,4], method = 'spearman')), 
                                         apply(InSilicoForeGround, 2, function(x) cor(x, InSilicoLandMarks[,5], method = 'spearman')), 
                                         apply(InSilicoForeGround, 2, function(x) cor(x, InSilicoLandMarks[,6], method = 'spearman')))
cell.info = bamfile.table[which(paste0(bamfile.table[,1], ".bam") %in% colnames(InSilicoForeGround)), 2]
library(gplots) 
library(RColorBrewer)
library(devtools)
source_url("https://raw.githubusercontent.com/obigriffith/biostar-tutorials/master/Heatmaps/heatmap.3.R")
scalered <- colorRampPalette(c("white", "red"), space = "rgb")(256)
rcols1 = brewer.pal(6, "Accent")[1:6]
rowcols1 = rcols1[cell.info]
rcols2 = brewer.pal(6, "Dark2")[1:6]
rowcols2 = rcols2[InSilicoLandMarkAssignments]
rowcols = rbind(rowcols1, rowcols2)
rownames(rowcols) = c("rep info", "cluster")
heatmap.3(InSilicoCell2LandmarkCorrelation, dendrogram='none', Rowv=FALSE, Colv=FALSE,
          trace='none', col = scalered, margin = c(5, 5), density.info = "none", 
          RowSideColors = rowcols, RowSideColorsSize=2, symm=F,symkey=F,
          symbreaks=F, scale="none")
legend("bottomleft", legend = c(unique(cell.info), paste0("cluster ", 1:6)), col = c(rcols1, rcols2), border=FALSE, bty="n", y.intersp = 0.7, cex=0.7, pch = 15)
heatmap.2(InSilicoCell2LandmarkCorrelation, dendrogram='none', Rowv=FALSE, 
          Colv=FALSE, trace='none', col = scalered, margin = c(5, 5), 
          density.info = "none", RowSideColors = rowcols1, symm=F,
          symkey=F,symbreaks=F, scale="none")
legend("bottomleft", legend = c(unique(cell.info)), col = rcols1, border=FALSE, 
       bty="n", y.intersp = 0.7, cex=0.7, pch = 15)
```

To view the above figure better, we're normalize the correlations of each cell so that the maximum is 1.

```{r cache=TRUE}
# for prettyness
for(i in 1:dim(InSilicoCell2LandmarkCorrelation)[1]){
  InSilicoCell2LandmarkCorrelation[i,] = InSilicoCell2LandmarkCorrelation[i,]/mean(InSilicoCell2LandmarkCorrelation[i,])
}
library(gplots) 
library(RColorBrewer)
heatmap.2(InSilicoCell2LandmarkCorrelation, dendrogram='none', Rowv=FALSE, Colv=FALSE, trace='none', col = scalered, margin = c(5, 5), density.info = "none", 
          RowSideColors = rowcols1, symm=F,symkey=F,symbreaks=F, scale="none")
legend("bottomleft", legend = c(unique(cell.info)), col = rcols1, border=FALSE, 
       bty="n", y.intersp = 0.7, cex=0.7, pch = 15)
```

The clustering correctly classifies all cells except for 1.  This assumes we knew the number of clusters.  What if we let scABC choose the number of clusters?

```{r message=FALSE, cache=TRUE}
#
InSilicoBackGroundMedian = apply(InSilicoBackGround, 2, median)
InSilicoGapStat = getGapStat(InSilicoForeGround, InSilicoBackGroundMedian, 
                             nClusters=1:10, nPerm = 20, quiet = TRUE)
InSilicoGapStat$nClusterOptimal
plotGapStat(InSilicoGapStat, nClusters = 1:10, main = "In Silico Gap Stat")
```

scABC chooses 6 clusters and we can continue with this number of clusters.

```{r cache=TRUE}
#
InSilicoPeakSelection = getClusterSpecificPvalue(ForeGround=InSilicoForeGround, cluster_assignments = InSilicoLandMarkAssignments, background_medians = InSilicoBackGroundMedian)
InSilicoPeakPvals = InSilicoPeakSelection$pvalue
head(InSilicoPeakPvals)
```

Let's visualize the data.

```{r cache=TRUE}
# all peaks 
scalered <- colorRampPalette(c("white", "red"), space = "rgb")(256)
rcols1 = brewer.pal(6, "Accent")[1:6]
rowcols1 = rcols1[cell.info]
rcols2 = brewer.pal(6, "Dark2")[1:6]
rowcols2 = rcols2[InSilicoLandMarkAssignments]
rowcols = rbind(rowcols1, rowcols2)
rownames(rowcols) = c("rep info", "cluster")
x = t(InSilicoForeGround)
x = x[ ,head(order(rowSums(InSilicoForeGround), decreasing = TRUE), 10000)]
d = as.dist(1 - cor(x, method = "spearman"))
col.clus = hclust(d, method ="average")
#truncate to see patterns
x[which(x > 5)] = 5
heatmap.3(x, dendrogram='none', Rowv=FALSE, Colv=col.clus, trace='none', col = scalered, 
          margin = c(5, 5), density.info = "none", RowSideColors = rowcols, 
          RowSideColorsSize=2, main = "All peaks")
legend("bottomleft", legend = c(unique(cell.info), paste0("cluster ", 1:6)), 
       col = c(rcols1, rcols2), border=FALSE, bty="n", y.intersp = 0.7, cex=0.7, pch = 15)
```

We'll select the cluster specific peaks by selecting the 10,000 peaks with the smallest p-values across all clusters.  
```{r cache=TRUE}
# cluster specific peaks
ClusterSpecificPeaks = head(order(apply(InSilicoPeakPvals, 1, min), 
                                  decreasing = FALSE), 10000)
length(ClusterSpecificPeaks)
x = t(InSilicoForeGround[ClusterSpecificPeaks, ])
d = as.dist(1 - cor(x, method = "spearman"))
col.clus = hclust(d, method ="average")
#truncate to see patterns
x[which(x > 5)] = 5
heatmap.3(x, dendrogram='none', Rowv=FALSE, Colv=as.dendrogram(col.clus), trace='none', 
          col = scalered, margin = c(5, 5), density.info = "none", RowSideColors = rowcols,
          RowSideColorsSize=2, main = "Cluster specific peaks")
legend("bottomleft", legend = c(unique(cell.info), paste0("cluster ", 1:6)), 
       col = c(rcols1, rcols2), border=FALSE, bty="n", y.intersp = 0.7, cex=0.7, pch = 15)
```

Instead of all of the above commands to obtain the clustered and processed data, we can simply use the scABC command to do all of the above.

```{r cache=TRUE}
library(GenomicRanges)
library(Rsamtools)
library(WeightedCluster)
peakfile = "NoTreatment/6Lines/mergeAll.tn5.pf.gappedPeak"
bamfile.table = read.table(file = "NoTreatment/6Lines/SRX&Type&Batch.txt")
bamfiles = paste0("bams/", bamfile.table[,1], ".bam")
# main function
InSilicoSCABC = scABC(bamfiles, peakfile)
```

We'll subset a selection of peaks to look at top 10,000 cluster specific peaks.
```{r cache=TRUE}
#
ClusterSpecificPeaksV2 = head(order(apply(InSilicoSCABC$PeakPVals, 1, min), 
                                    decreasing = FALSE), 10000)
length(intersect(ClusterSpecificPeaks, ClusterSpecificPeaksV2))
```

```{r cache=TRUE}
scalered <- colorRampPalette(c("white", "red"), space = "rgb")(256)
rcols1 = brewer.pal(6, "Accent")[1:6]
rowcols1 = rcols1[cell.info]
rcols2 = brewer.pal(6, "Dark2")[1:6]
rowcols2 = rcols2[InSilicoSCABC$cluster_assignments]
rowcols = rbind(rowcols1, rowcols2)
rownames(rowcols) = c("rep info", "cluster")
x = t(InSilicoSCABC$ForeGroundMatrix[ClusterSpecificPeaksV2, ])
d = as.dist(1 - cor(x, method = "spearman"))
col.clus = hclust(d, method ="average")
x[which(x > 5)] = 5
heatmap.3(x, dendrogram='none', Rowv=FALSE, Colv=as.dendrogram(col.clus), trace='none', 
          col = scalered, margin = c(5, 5), density.info = "none", RowSideColors = rowcols,
          RowSideColorsSize=2, main = "Cluster specific peaks")
legend("bottomleft", legend = c(unique(cell.info), paste0("cluster ", 1:6)), 
       col = c(rcols1, rcols2), border=FALSE, bty="n", y.intersp = 0.7, cex=0.7, pch = 15)
```

We can also select cluster specific peaks by taking peaks that are less than some threshhold.  We have to set it to be really small because Rmarkdown does not support large matrices.  

```{r cache=TRUE}
#
ClusterSpecificPeaksV3 = which(apply(InSilicoSCABC$PeakPVals, 1, min) < 0.0001)
length(ClusterSpecificPeaksV3)
```

```{r cache=TRUE}
# all peaks are cluster specific
scalered <- colorRampPalette(c("white", "red"), space = "rgb")(256)
rcols1 = brewer.pal(6, "Accent")[1:6]
rowcols1 = rcols1[cell.info]
rcols2 = brewer.pal(6, "Dark2")[1:6]
rowcols2 = rcols2[InSilicoSCABC$cluster_assignments]
rowcols = rbind(rowcols1, rowcols2)
rownames(rowcols) = c("rep info", "cluster")
x = t(InSilicoSCABC$ForeGroundMatrix[ClusterSpecificPeaksV3, ])
d = as.dist(1 - cor(x, method = "spearman"))
col.clus = hclust(d, method ="average")
x[which(x > 5)] = 5
heatmap.3(x, dendrogram='none', Rowv=FALSE, Colv=as.dendrogram(col.clus), 
          trace='none', col = scalered, margin = c(5, 5), density.info = "none", 
          RowSideColors = rowcols, RowSideColorsSize=2, main = "Cluster specific peaks")
legend("bottomleft", legend = c(unique(cell.info), paste0("cluster ", 1:6)), 
       col = c(rcols1, rcols2), border=FALSE, bty="n", y.intersp = 0.7, cex=0.7, pch = 15)
```

# Using chromVAR on cluster specific peaks

To identify transcription factors active in cluster specific peaks we'll use chromVAR (https://greenleaflab.github.io/chromVAR/) to compute TF motifs. 

First we'll look at the naive application of chromVAR without selecting for cluster specific peaks

```{r cache=TRUE, message=FALSE}
#
ClusterSpecificPeaks = data.frame(chrom = InSilicoSCABC$peaks$chrom, start = InSilicoSCABC$peaks$start, end = InSilicoSCABC$peaks$end, cluster1pvals = InSilicoSCABC$PeakPVals[,1], cluster2pvals = InSilicoSCABC$PeakPVals[,2], cluster3pvals = InSilicoSCABC$PeakPVals[,3], cluster4pvals = InSilicoSCABC$PeakPVals[,4], cluster5pvals = InSilicoSCABC$PeakPVals[,5], cluster6pvals = InSilicoSCABC$PeakPVals[,6])
dim(ClusterSpecificPeaks)
write.table(ClusterSpecificPeaks, file = "InSilicoPeaksWithPvals.txt", sep = "\t", quote = FALSE, col.names = FALSE, row.names = FALSE)
#BiocInstaller::biocLite("GreenleafLab/chromVAR")
library(chromVAR)
library(motifmatchr)
library(Matrix)
library(SummarizedExperiment)
extra_cols = c(4, 5, 6, 7, 8, 9)
names(extra_cols) = c("cluster1", "cluster2", "cluster3", "cluster4", "cluster5", "cluster6")
peaks = getPeaks("~/scRNAseqAnalysis/scATAC-RNAseq/scABC/vignettes/InSilicoPeaksWithPvals.txt", extra_cols = extra_cols)
# peaks include broad peaks, remove broad peaks
hist(width(peaks), breaks = 200)
length(peaks)
peaks = peaks[which(width(peaks) <= 800)]
length(peaks)
peaks = resize(peaks, width = 500, fix = "center")
peaks = sort(peaks)
bamfiles = paste0("bams/", colnames(InSilicoSCABC$ForeGroundMatrix))
frag_counts = getCounts(bamfiles, peaks, paired = FALSE, 
                         colData = data.frame(cluster.assigment = factor(InSilicoSCABC$cluster_assignments)))
library(BSgenome.Hsapiens.UCSC.hg19)
frag_counts = addGCBias(frag_counts, genome = BSgenome.Hsapiens.UCSC.hg19)
motifs <- getJasparMotifs()
motifs.matched = matchMotifs(motifs, frag_counts, genome = BSgenome.Hsapiens.UCSC.hg19)
dev = computeDeviations(object = frag_counts, annotations = motifs.matched)
dev.scores = deviationScores(dev)
variability = computeVariability(dev)
plotVariability(variability, use_plotly = TRUE) 
variability[head(order(variability$variability, decreasing = TRUE), 20), ]
tsne = deviationsTsne(dev, threshold = 1, perplexity = 10)
tsne_plot = plotDeviationsTsne(dev, tsne, annotation = "FOSL2", 
                                 sample_column = "cluster.assigment", shiny = FALSE)
tsne_plot[[1]]
tsne_plot[[2]]
tsne_plot = plotDeviationsTsne(dev, tsne, annotation = "CTCF", 
                                 sample_column = "cluster.assigment", shiny = FALSE)
tsne_plot[[2]]
tsne_plot = plotDeviationsTsne(dev, tsne, annotation = "GATA1::TAL1", 
                                 sample_column = "cluster.assigment", shiny = FALSE)
tsne_plot[[2]]
top_motifs = variability$name[head(order(variability$variability, decreasing = TRUE), 20)]
names(top_motifs) = rownames(variability[head(order(variability$variability, decreasing = TRUE), 20), ])
top_devs = dev.scores[which(rownames(dev.scores) %in% names(top_motifs)), ]
rownames(top_devs) = top_motifs[match(rownames(top_devs), names(top_motifs))]
library(gplots) 
library(RColorBrewer)
scalebluered <- colorRampPalette(c("blue", "white", "red"), space = "rgb")(256)
cols = brewer.pal(6, "Dark2")[1:6]
cols = cols[c(6, 1, 4, 2, 5, 3)] # reorder to agree with Zhi's figure
rowcols = cols[InSilicoSCABC$cluster_assignments]
library(vegan)
d = vegdist(top_devs, method = "euclidean")
col.clus = hclust(d, "centroid")
heatmap.2(t(top_devs), dendrogram='none', Rowv=FALSE, Colv=as.dendrogram(col.clus), trace='none', col = scalebluered, density.info = "none", RowSideColors = rowcols, margin = c(10, 1))
```

Most of the TFs are BJ specific. Let's look at when we take cluster specific peaks, but an equal number of each cluster.

```{r cache=TRUE, message=FALSE}
topPeaks = peaks[which(apply(elementMetadata(peaks), 1, min) < 1e-5)]
fragCountsTopPeaks = getCounts(bamfiles, topPeaks, paired = FALSE, 
                         colData = data.frame(cluster.assigment = factor(InSilicoSCABC$cluster_assignments)))
fragCountsTopPeaks = addGCBias(fragCountsTopPeaks, genome = BSgenome.Hsapiens.UCSC.hg19)
motifs <- getJasparMotifs()
motifsTopPeaks.matched = matchMotifs(motifs, fragCountsTopPeaks, genome = BSgenome.Hsapiens.UCSC.hg19)
devTopPeaks = computeDeviations(object = fragCountsTopPeaks, annotations = motifsTopPeaks.matched)
devTopPeaks.scores = deviationScores(devTopPeaks)
variabilityTopPeaks = computeVariability(devTopPeaks)
plotVariability(variabilityTopPeaks, use_plotly = TRUE) 
variabilityTopPeaks[head(order(variabilityTopPeaks$variability, decreasing = TRUE), 20), ]
intersect(variabilityTopPeaks$name[head(order(variabilityTopPeaks$variability, decreasing = TRUE), 20)], variability$name[head(order(variability$variability, decreasing = TRUE), 20)])

tsneTopPeaks = deviationsTsne(devTopPeaks, threshold = 1, perplexity = 10)
tsneTopPeaks_plot = plotDeviationsTsne(devTopPeaks, tsneTopPeaks, annotation = "FOSL2", 
                                 sample_column = "cluster.assigment", shiny = FALSE)
tsne_plot[[1]]
```

A large number intersect, which is good.

```{r cache=TRUE}
topMotifsTopPeaks = variabilityTopPeaks$name[head(order(variabilityTopPeaks$variability, decreasing = TRUE), 20)]
names(topMotifsTopPeaks) = rownames(variabilityTopPeaks[head(order(variabilityTopPeaks$variability, decreasing = TRUE), 20), ])
topDevsTopPeaks = devTopPeaks.scores[which(rownames(devTopPeaks.scores) %in% names(topMotifsTopPeaks)), ]
rownames(topDevsTopPeaks) = topMotifsTopPeaks[match(rownames(topDevsTopPeaks), names(topMotifsTopPeaks))]
topDevsTopPeaks[which(is.na(topDevsTopPeaks))] = 0
library(gplots) 
library(RColorBrewer)
scalebluered <- colorRampPalette(c("blue", "white", "red"), space = "rgb")(256)
cols = brewer.pal(6, "Dark2")[1:6]
cols = cols[c(6, 1, 4, 2, 5, 3)] # reorder to agree with Zhi's figure
rowcols = cols[InSilicoSCABC$cluster_assignments]
library(vegan)
d = vegdist(topDevsTopPeaks, method = "euclidean")
col.clus = hclust(d, "centroid")
heatmap.2(t(topDevsTopPeaks), dendrogram='none', Rowv=FALSE, Colv=as.dendrogram(col.clus), trace='none', col = scalebluered, density.info = "none", RowSideColors = rowcols, margin = c(10, 1))
```

In the cluster specific peaks the Jun family motifs in BJ are still present, but now we see the Gata1-Tal1 motif in K562 cells (see https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3431495/ or http://mcb.asm.org/content/25/4/1215.full) and Pou motifs in H1 cells (espcially Pou3f2, aka Oct4).

# Session Info

```{r}
sessionInfo()
```